### DISCUSSION TITLE
USD Light Link

### DISCUSSION THREAD

**Discussion by clebercoutinho** [Autodesk/maya-usd#2851] [Q&A]
Hi,
It seems that Maya Light Link does not work with USD? is it something that I have to turn on?

It would be very very crazy to have a very complex system like USD, and not have a Light Link Relation working.

**Comment by santosg87** [comment]
Hi @clebercoutinho,

we have a basic light linking setup that USD itself uses - although mayaUSD light support is relatively low. we currently have only about 3 light types working with MayaUSD - DistantLight (maya equivalent: Directional Light), RecLight (maya equivalent: Area Light) and SphereLight (maya equivalent: Point Light).

I'm not 100% sure Maya light Linking is UFE aware, so this would require proper support to begin with. :)

**Reply by clebercoutinho** [reply]
Hey Santos,

Thanks for replying me!
So these 3 USD lights (DistantLight,RecLight, SphereLight) have not been implemented in Maya. and with Light link.
For the lighting department,  "lighting link" is crucial.

**Reply by santosg87** [reply]
Correct!
As I mentioned, in order for those to be working on the Maya side, it will require new support to be added on Maya itself.

I will bring the feedback to the rest of the team! :)

cheers!

**Reply by jbcastella** [reply]
Hey team, any news on these topic? we are starting a film production and we have to make the decision to start with USD, but if you continue with the limitation of the light linking it is impossible to do a proper assembling inside maya without it.

Best.

**Reply by clebercoutinho** [reply]
Houdini and Katana have usd implemented more than 5 years already and ->>>>light linking working with USD.
I would say that all USD setup in Maya is completely useless without a light link.
Therefore we are planning to move to Houdini if they don't update this so basic feature.

**Comment by wallworm** [comment]
Light linking is an essential feature on our roadmap. Exactly when we get to it is the only question at the moment. When we have this queued up for development, we will certainly share that news here.

**Comment by sepux** [comment]
Another vote for light linking - This is needed badly for Maya USD. This should be one of the priorities

About Houdini implementing USD more than 5 years ago that isn't true - Solaris came out H 18 which is basically at the end of 2019.  Obviously was in alpha/beta before that but it was not available for the public until H18. Probably the same with Katana (though I'm not sure)

**Reply by clebercoutinho** [reply]
Siggraph 2018 had a few studios showing USD being used in several shots. Even animal logic in maya. Pixar - katana. etc
A few studios get early access always. (houdini 17.0 started to get some USD stuff too)

yes! Light Link is a must :)

**Comment by Mudoglu** [comment]
Yes please give Light Linking a push!

**Comment by clebercoutinho** [comment]
one more year, one more update for USD.
no light link. :(

**Reply by Mudoglu** [reply]
There is a way with Bifrost. Do a search on Bifrost Discord.

**Reply by clebercoutinho** [reply]
hummmm< Mudoglu! I will try find, crossing fingers, thanks

**Reply by BigRoy** [reply]
Please link here. By the way, is Light Linking implemented in the USD API itself? If so, I'd be happy to prototype some Qt tools to do light linking. Please link that Bifrost thread/post here as well (Copy link from Discord)

**Reply by Mudoglu** [reply]
Here it is.
https://discord.com/channels/872260298508222534/1069220721642962975

**Reply by BigRoy** [reply]
Which discord channel is this? It seems that it isn't public enough for me to get there by just this link.

Nevermind, it's **Bifrost Addicts** - invite link here: https://forums.autodesk.com/t5/bifrost-forum/bifrost-addicts-has-migrated-to-discord/td-p/10710916

I was checking it via my wrong accounts.

**Reply by BigRoy** [reply]
Thanks - [here's a quick UI to do these actions outside](https://gist.github.com/BigRoy/1895108757a9122f9cb90abf08c70015) of Bifrost on your USD data via maya usd proxy shape. Select the lights + target objects in the outliner and click the buttons in the UI.

_Tested in Maya 2024.2 with Maya USD 0.26.0 and Arnold rendering_ this little tool worked to manage the light linking.

You could look into maybe using [`usd_qtpy`](https://github.com/BigRoy/usd-qtpy/tree/main/usd_qtpy) to debug the actual opinions made in the USD data in the prim spec editor.

I have no idea how many people are actively using Maya USD for rendering and how worthwhile it'd be to expanding this quick example more.

It shouldn't be too hard to e.g. have it update automatically on selected light/targets so you can see what objects are currently in the Light's Include or Exclude sets.

**Reply by clebercoutinho** [reply]
Hey BigRoy.
Great that you have some python to make some magic.
We would love use USD with rendering in Maya, but without Light Linking I would state as completely useless system. (USD+Bifrost+LookDev setup)

**Reply by clebercoutinho** [reply]
maya 2025 is out.
one more year - no light link for USD.
well done autodesk :)

**Reply by wallworm** [reply]
> maya 2025 is out. one more year - no light link for USD. well done autodesk :)

This is still on our roadmap. We just have not gotten to this particular item yet. It's more complex than just supporting light linking. If we keep it simply as USD prims linking to USD lights, then the path forward is simply exposing a UX for linking in a USD context. But as soon as you start talking mixing USD with Maya (linking a Maya object to a USD Light for example), then you run into problems. Same with linking a USD prim to a Maya light--the question becomes how to properly send that back into the pipeline. Right now I'm leaning to keeping it in USD Land (linking USD Prims to USD Lights) for simplicity. But what are the expectations of all our users on this topic?

**Reply by Mudoglu** [reply]
Yes keep it in USD land. Link USD Prims to USD Lights.

**Reply by clebercoutinho** [reply]
Hey,

We could keep all in USD land, and allow (render developers -> vray, redshift, etc) to pipe their lights as extra attributes in these USD lights.
This is the way Houdini works.
thanks

**Reply by scottrenaud** [reply]
Hey folks, 

We just posted some designs for this: [Feedback needed on new UI - Light Linking with Collections in the Attribute Editor](https://github.com/Autodesk/maya-usd/discussions/3816).

Please take a look and let us know what you think, thanks!
